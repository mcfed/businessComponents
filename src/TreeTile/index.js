import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {Tree, Card, Checkbox} from 'antd';

const TreeNode = Tree.TreeNode;

class TreeTile extends Component {
  constructor(props) {
    super(props);
    this.state = {
      indeterminate: false,
      checkAll: false,
      dataSourceKeys: [],
      checkedKeys: []
    };
  }

  componentDidMount() {
    let {dataSource} = this.props;
    let dataSourceKeys = this.getKeysFromMap(dataSource, []);
    this.setState({
      dataSourceKeys
    });

    let {checkedKeys} = this.props;
    if (checkedKeys instanceof Array) {
      this.setState({
        checkedKeys
      });
    } else {
      this.setState({
        checkedKeys: []
      });
    }
  }

  getKeysFromMap = (data, dataSourceKeys) => {
    data.map(d => {
      if (!!d.children) {
        this.getKeysFromMap(d.children, dataSourceKeys);
      }
      dataSourceKeys.push(d.key);
    });
    return dataSourceKeys;
  };

  renderTreeNodes = data =>
    data.map(item => {
      if (item.children) {
        return (
          <TreeNode title={item.title} key={item.key} dataRef={item}>
            {this.renderTreeNodes(item.children)}
          </TreeNode>
        );
      }
      return <TreeNode {...item} />;
    });

  onCheckAll = e => {
    let {dataSourceKeys} = this.state;
    let checkedKeys = e.target.checked ? dataSourceKeys : [];
    this.setState(
      {
        checkedKeys: checkedKeys,
        indeterminate: false,
        checkAll: e.target.checked
      },
      () => {
        this.props.onChange(checkedKeys);
      }
    );
  };

  onCheck = checkedKeys => {
    let {dataSourceKeys} = this.state;
    this.setState(
      {
        checkedKeys: checkedKeys,
        indeterminate:
          !!checkedKeys &&
          checkedKeys.length !== 0 &&
          checkedKeys.length < dataSourceKeys.length,
        checkAll: !!checkedKeys && checkedKeys.length === dataSourceKeys.length
      },
      () => {
        this.props.onChange(checkedKeys);
      }
    );
  };

  render() {
    const {title, dataSource} = this.props;
    let {checkedKeys} = this.state;
    return (
      <Card
        size='small'
        title={title}
        extra={
          <Checkbox
            onChange={this.onCheckAll}
            indeterminate={this.state.indeterminate}
            checked={this.state.checkAll}>
            全选
          </Checkbox>
        }
        style={{width: '100%'}}>
        {!!dataSource && dataSource.length > 0 ? (
          <Tree
            className='treeTile'
            name='tree'
            checkable
            blockNode={false}
            defaultExpandAll
            onCheck={this.onCheck}
            checkedKeys={checkedKeys}>
            {this.renderTreeNodes(dataSource)}
          </Tree>
        ) : (
          <p className='treeNoData'>no data</p>
        )}
      </Card>
    );
  }
}

TreeTile.propTypes = {
  /**
   * 标题
   */
  title: PropTypes.string,
  /**
   * （受控）选中复选框的树节点（注意：父子节点有关联，如果传入父节点 key，则子节点自动选中；相应当子节点 key 都传入，父节点也自动选中。当设置 checkable 和 checkStrictly，它是*一个有checked和halfChecked属性的对象，并且父子节点的选中与否不再关联
   */
  checkedKeys: PropTypes.array,
  /**
   * 数据源
   */
  dataSource: PropTypes.array.isRequired
};

TreeTile.defaultProps = {
  title: '请选择',
  checkedKeys: [],
  dataSource: []
};

export default TreeTile;
