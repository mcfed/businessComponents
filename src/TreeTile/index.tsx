import React from 'react';
import {Tree, Card, Checkbox} from 'antd';
const TreeNode = Tree.TreeNode;

interface TreeTileProps {
  title?: string;
  checkedKeys?: checkedKeysType;
  dataSource: any;
  onChange?: (arg: any) => void;
}

interface TreeTileStates {
  indeterminate: boolean;
  checkAll: boolean;
  dataSourceKeys: any;
  checkedKeys: checkedKeysType;
}

type checkedKeysType = checkedKeysArrayType | checkedKeysObjectType;

type checkedKeysArrayType = string[];
type checkedKeysObjectType = {
  checked: string[];
  halfChecked: string[];
};

class TreeTile extends React.Component<TreeTileProps, TreeTileStates> {
  static defaultProps = {
    dataSource: [],
    title: '请选择'
  };

  constructor(props: TreeTileProps) {
    super(props);
    this.state = {
      indeterminate: false,
      checkAll: false,
      dataSourceKeys: [],
      checkedKeys: []
    };
  }

  componentDidMount() {
    let {dataSource} = this.props;
    let dataSourceKeys = this.getKeysFromMap(dataSource, []);
    this.setState({
      dataSourceKeys
    });

    let {checkedKeys} = this.props;
    if (checkedKeys instanceof Array) {
      this.setState({
        checkedKeys
      });
    } else {
      this.setState({
        checkedKeys: []
      });
    }
  }

  getKeysFromMap = (data: any, dataSourceKeys: any) => {
    data.map((d: any) => {
      if (!!d.children) {
        this.getKeysFromMap(d.children, dataSourceKeys);
      }
      dataSourceKeys.push(d.key);
    });
    return dataSourceKeys;
  };

  renderTreeNodes = (data: any) =>
    data.map((item: any) => {
      if (item.children) {
        return (
          <TreeNode title={item.title} key={item.key} dataRef={item}>
            {this.renderTreeNodes(item.children)}
          </TreeNode>
        );
      }
      return <TreeNode {...item} />;
    });

  onChange = (item: any) => {
    const {onChange} = this.props;
    if (!onChange) return null;
    return onChange(item);
  };

  onCheckAll = (e: any) => {
    let {dataSourceKeys} = this.state;
    let checkedKeys = e.target.checked ? dataSourceKeys : [];
    this.setState(
      {
        checkedKeys: checkedKeys,
        indeterminate: false,
        checkAll: e.target.checked
      },
      () => {
        this.onChange(checkedKeys);
      }
    );
  };

  isCheckedKeysArrayType = (
    keys: checkedKeysType
  ): keys is checkedKeysArrayType => {
    return keys instanceof Array;
  };

  onCheck = (checkedKeys: checkedKeysType) => {
    // istanbul ignore else
    if (this.isCheckedKeysArrayType(checkedKeys)) {
      let {dataSourceKeys} = this.state;
      this.setState(
        {
          checkedKeys: checkedKeys,
          indeterminate:
            !!checkedKeys &&
            checkedKeys.length !== 0 &&
            checkedKeys.length < dataSourceKeys.length,
          checkAll:
            !!checkedKeys && checkedKeys.length === dataSourceKeys.length
        },
        () => {
          this.onChange(checkedKeys);
        }
      );
    }
  };

  render() {
    const {title, dataSource} = this.props;
    let {checkedKeys} = this.state;
    return (
      //@ts-ignore
      <Card
        // @ts-ignore: 自定义属性，忽略检查
        size='small'
        title={title}
        extra={
          <Checkbox
            onChange={this.onCheckAll}
            indeterminate={this.state.indeterminate}
            checked={this.state.checkAll}>
            全选
          </Checkbox>
        }
        style={{width: '100%'}}>
        {!!dataSource && dataSource.length > 0 ? (
          //@ts-ignore
          <Tree
            className='treeTile'
            // @ts-ignore: 自定义属性，忽略检查
            name='tree'
            checkable
            blockNode={false}
            defaultExpandAll
            onCheck={this.onCheck}
            checkedKeys={checkedKeys}>
            {this.renderTreeNodes(dataSource)}
          </Tree>
        ) : (
          <p className='treeNoData'>no data</p>
        )}
      </Card>
    );
  }
}

export default TreeTile;
