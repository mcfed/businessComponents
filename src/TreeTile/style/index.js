import 'antd/es/tree/style';
import 'antd/es/card/style';
import 'antd/es/checkbox/style';

import './style.less';
