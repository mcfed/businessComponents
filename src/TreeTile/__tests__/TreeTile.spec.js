import {shallow, mount, render} from 'enzyme';
import React from 'react';
import TreeTile from '../index.tsx';

const setup = props => {
  const onChange = item => <div>{item}</div>;
  const wrapper = shallow(<TreeTile {...props} />);
  return {
    wrapper,
    props
  };
};

describe('快照测试', () => {
  it('TreeTile 快照测试', () => {
    const {wrapper} = setup();
    expect(wrapper).toMatchSnapshot();
  });
});

describe.skip('TreeTile render 是否正确渲染', () => {
  const treeData = [
      {
        title: 'Node1',
        value: '0-0',
        key: '0-0',
        children: [
          {
            title: 'Child Node1',
            value: '0-0-0',
            key: '0-0-0'
          }
        ]
      },
      {
        title: 'Node2',
        value: '0-1',
        key: '0-1',
        children: [
          {
            title: 'Child Node3',
            value: '0-1-0',
            key: '0-1-0'
          },
          {
            title: 'Child Node4',
            value: '0-1-1',
            key: '0-1-1'
          },
          {
            title: 'Child Node5',
            value: '0-1-2',
            key: '0-1-2'
          }
        ]
      }
    ],
    checkedKeys = ['0-1-2'];
  const defaultProps = {
    title: '测试用例',
    checkedKeys: checkedKeys,
    dataSource: treeData
  };
  it('TreeTile 不传 title 参数时渲染是否正确', () => {
    const {wrapper, props} = setup({});
    expect(wrapper.find('Card').prop('title')).toBe('请选择');
  });

  it('TreeTile 不传 dataSource 时渲染是否正确', () => {
    const {wrapper, props} = setup({});
    expect(wrapper.find('Tree').length).toBe(0);
  });

  it('TreeTile 不传 checkedKeys 时渲染是否正确', () => {
    const {wrapper, props} = setup({});
    expect(wrapper.instance().state.checkedKeys).toEqual([]);
  });

  it('TreeTile 传入 title 参数时渲染是否正确', () => {
    const {wrapper, props} = setup({...defaultProps});
    expect(wrapper.find('Card').prop('title')).toBe('测试用例');
  });

  it('TreeTile 传入 dataSource 参数时渲染是否正确', () => {
    const {wrapper, props} = setup({...defaultProps});
    expect(wrapper.find('Tree').length).toBe(1);
  });

  it('TreeTile 传入 dataSource 参数后节点个数渲染是否正确', () => {
    const {wrapper, props} = setup({...defaultProps});
    expect(wrapper.find('TreeNode').length).toBe(6);
  });

  it('TreeTile 传入 checkedKeys 参数后选中状态渲染是否正确', () => {
    const {wrapper, props} = setup({...defaultProps});
    expect(wrapper.instance().state.checkedKeys).toEqual(checkedKeys);
  });
});

describe.skip('TreeTile 方法测试', () => {
  const treeData = [
      {
        title: 'Node1',
        value: '0-0',
        key: '0-0',
        children: [
          {
            title: 'Child Node1',
            value: '0-0-0',
            key: '0-0-0'
          }
        ]
      },
      {
        title: 'Node2',
        value: '0-1',
        key: '0-1',
        children: [
          {
            title: 'Child Node3',
            value: '0-1-0',
            key: '0-1-0'
          },
          {
            title: 'Child Node4',
            value: '0-1-1',
            key: '0-1-1'
          },
          {
            title: 'Child Node5',
            value: '0-1-2',
            key: '0-1-2'
          }
        ]
      }
    ],
    checkedKeys = [];
  const onChange = jest.fn();
  const defaultProps = {
    title: '测试用例',
    checkedKeys: checkedKeys,
    dataSource: treeData,
    onChange: onChange
  };

  it('getKeysFromMap 方法调用', () => {
    const {wrapper, props} = setup({...defaultProps});
    wrapper.instance().getKeysFromMap(treeData, checkedKeys);
    expect(wrapper.instance().state.dataSourceKeys).toEqual(checkedKeys);
  });

  it('renderTreeNodes 方法调用', () => {
    const {wrapper, props} = setup({...defaultProps});
    wrapper.instance().renderTreeNodes(treeData);
    expect(wrapper.find('TreeNode').length).toBe(6);
  });

  it('onCheckAll 方法调用', () => {
    const {wrapper, props} = setup({...defaultProps});
    const e = {target: {checked: true}};
    wrapper.instance().onCheckAll(e);
    expect(wrapper.instance().state.checkedKeys).toEqual(checkedKeys);
    expect(wrapper.instance().state.indeterminate).toEqual(false);
    expect(onChange).toBeCalled();
  });

  it('onCheck 方法调用', () => {
    const {wrapper, props} = setup({...defaultProps});
    wrapper.instance().onCheck(checkedKeys);
    wrapper.instance().state.dataSourceKeys = checkedKeys;
    expect(wrapper.instance().state.checkedKeys).toEqual(checkedKeys);
    expect(wrapper.instance().state.indeterminate).toEqual(false);
    expect(wrapper.instance().state.checkAll).toEqual(true);
    expect(onChange).toBeCalled();
  });
});
