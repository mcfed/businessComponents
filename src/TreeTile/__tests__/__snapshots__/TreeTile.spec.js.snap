// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`快照测试 TreeTile 快照测试 1`] = `
<Card
  extra={
    <Checkbox
      checked={false}
      indeterminate={false}
      onChange={[Function]}
    >
      全选
    </Checkbox>
  }
  size="small"
  style={
    Object {
      "width": "100%",
    }
  }
  title="请选择"
>
  <p
    className="treeNoData"
  >
    no data
  </p>
</Card>
`;
